fml/platform/fuchsia/message_loop_fuchsia.cc

```diff
@@ -10,29 +10,46 @@
 
 namespace fml {
 
-MessageLoopFuchsia::MessageLoopFuchsia()
-    : loop_(&kAsyncLoopConfigAttachToCurrentThread) {}
+MessageLoopFuchsia::MessageLoopFuchsia() : running_(false) {
+  auto status = zx_timer_create(0, ZX_CLOCK_MONOTONIC, &timer_);
+  FML_CHECK(status == ZX_OK);
+}
 
 MessageLoopFuchsia::~MessageLoopFuchsia() = default;
 
 void MessageLoopFuchsia::Run() {
-  loop_.Run();
+  zx_signals_t observed;
+  running_ = true;
+
+  while (running_) {
+    auto status = zx_object_wait_one(timer_, ZX_TIMER_SIGNALED,
```

Comment:
Communication over FIDL requires a multiplexed wait over many handles and the presence of a loop is assumed. We cannot have a blocking wait on a single handle here as it would disable all FIDL communication on engine managed threads.

I suggest creating an async_loop_t as well as a timer in this class and run the loop here (with the timer scheduled). The wait for the timer to be signaled cannot be blocking as a non-multiplexing wait would disable FIDL communication.

Reply:
I suggest looking at one of the existing impls for prior art on the process of using a loop and an FD for a non-blocking multiplexed wait (message_loop_android.cc seems fine to copy concepts from).

Reply:
Agree with everything chinmay says; it sounds like you guys solved it offline though